xlsynth-g8r/src/equiv/easy_smt_backend.rs

```diff
@@ -574,10 +650,112 @@ impl Solver for EasySMTSolver {
     fn render(&mut self, bit_vec: &BitVec<SExpr>) -> String {
         let shared = Arc::clone(&self.context);
         let context = shared.lock().unwrap();
-        match bit_vec.get_rep() {
-            None => "<zero-width>".to_string(),
-            Some(rep) => context.display(rep.clone()).to_string(),
+
+        // Handle zero-width values first.
+        let rep = match bit_vec.get_rep() {
+            None => return "<zero-width>".to_string(),
+            Some(r) => r.clone(),
+        };
+
+        // We can only build the pretty `let` representation if the final
+        // expression is an atom that refers to an entry in our term cache.
+        let final_name_opt = context.get_atom(rep);
+        let final_name = match final_name_opt {
+            Some(name) if self.term_cache.contains_key(name) => name.to_string(),
+            // Fallback: display the raw S-expr as before.
+            _ => return context.display(rep).to_string(),
+        };
+
+        // ------------------------------------------------------------------
+        // Collect all term-cache names that the final expression transitively
+        // depends on.  We perform a DFS and record the names in post-order so
+        // that we obtain a topological ordering (dependencies before uses).
+        // ------------------------------------------------------------------
+
+        // Small helper to recursively gather atom names referenced in an SExpr.
+        fn gather_names(
+            expr: SExpr,
+            ctx: &Context,
+            term_cache: &HashMap<String, SExpr>,
+            out: &mut Vec<String>,
+        ) {
+            if expr.is_atom() {
+                if let Some(atom) = ctx.get_atom(expr) {
+                    if term_cache.contains_key(atom) {
+                        out.push(atom.to_string());
+                    }
+                }
+            } else if expr.is_list() {
+                if let Some(children) = ctx.get_list(expr) {
+                    for &child in children {
+                        gather_names(child, ctx, term_cache, out);
+                    }
+                }
+            }
         }
+
+        // Recursive DFS for topological sort.
+        fn dfs(
```

Comment:
I think you want a worklist based DFS because our programs will get large enough to overflow the stack.